internal/controllers/catalog_test.go

```diff
@@ -0,0 +1,97 @@
+package controllers_test
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/operator-framework/operator-controller/internal/controllers"
+
+	catalogd "github.com/operator-framework/catalogd/pkg/apis/core/v1beta1"
+	operatorsv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var _ = Describe("Operator Controller Catalog Reconciler", func() {
+	When("a catalog changes on cluster", func() {
+		var reconcileReqs, opNames []string
+		var cancel context.CancelFunc
+		var logCount int
+		var ctx context.Context
+		BeforeEach(func() {
+			ctx = context.Background()
+			mgr, err := ctrl.NewManager(cfg, ctrl.Options{Scheme: sch})
+			Expect(err).To(BeNil())
+
+			var fakeReconciler reconcile.Func = func(_ context.Context, request ctrl.Request) (ctrl.Result, error) {
+				reconcileReqs = append(reconcileReqs, request.Name)
+				return ctrl.Result{}, nil
+			}
+
+			err = controllers.SetupWithManager(fakeReconciler, mgr)
+			Expect(err).To(BeNil())
+
+			var mgrCtx context.Context
+			mgrCtx, cancel = context.WithCancel(ctx)
+
+			go func() {
+				err := mgr.Start(mgrCtx)
+				Expect(err).To(BeNil())
+			}()
+
+			opNames = []string{"prometheus", "project-quay"}
+			for _, p := range opNames {
+				op := &operatorsv1alpha1.Operator{ObjectMeta: metav1.ObjectMeta{Name: p}, Spec: operatorsv1alpha1.OperatorSpec{PackageName: p}}
+				err := cl.Create(ctx, op)
+				Expect(err).To(BeNil())
+			}
+			Eventually(func(g Gomega) {
+				By("verifying initial reconcile logs for operator creation")
+				g.Expect(len(reconcileReqs) >= len(opNames)).To(BeTrue())
+				g.Expect(reconcileReqs).To(ContainElements(opNames))
+				logCount = len(reconcileReqs)
+			}).Should(Succeed())
```

Comment:
Not positive this works, but docs make it seem like it should:
```suggestion
			Eventually(reconcileReqs).Should(Equal(opNames))
```
`Equal` instead of `ContainElements` because (I'm pretty sure) we can expect a specific order of reconciles since we don't currently configure concurrent reconciles. 

For the `logCount` thing, I wonder if there is an alternative that lets us:
- not need it
- increase clarity of the subsequent test assertions

Perhaps we could "reset" the reconciler (i.e. make the slice 0 length) after each assertion? Might need a mutex to avoid a race condition.

But if we did that, then all of the `Eventually` checks would (I think?) look like the suggestion in this comment.

Reply:
I think we should consider the possibility of reconciles happening out of order, especially if we're introducing a potential flake.

Reply:
Ah, yeah. That makes sense. I think the handler make a `List` call and that returns Operators in a non-deterministic order.

I checked out this PR locally to continue investigating the channel idea. I ran these updated tests with some extra debug logging and noticed that there are specs in operator_controller_test that don't clean up after themselves and are causing _these_ tests to have to deal with that pollution.

I have a fix for that here: #257 

I think we should merge that, rebase this, and then see again if we can improve on the way we assert for reconciles. I still think there's a way to do this with channels (perhaps receive from channel into a slice and then assert on the slice, but without the slice being reused for multiple assertions?).